### PR TITLE
[FEM] Same Optimizations from PR#4759 + Race Condition Fix

### PR DESCRIPTION
**Pull request by AjinkyaDahale**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

---
Possibly fixes some race conditions introduced in PR #4759.

Rebased and made a new draft instead of making a PR in that branch because it wasn't compiling without a rebase.

### CODE REVIEW

**Comment by AjinkyaDahale** [comment]
Forced push to commit adding thread-safe version instead of using `#pragma omp critical`

**Comment by berndhahnebach** [comment]
would you rebase the PR to make it run on CI?

**Comment by berndhahnebach** [comment]
https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_4985

**Comment by berndhahnebach** [comment]
Following a link to the branch on the CI-repository:

https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_4985

The CI-status is available on the latest commit of the branch.
If there is no status available the PR should be rebased on latest master.
Check pipeline branches to see if your PR has been run by the CI.

https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches

**Comment by AjinkyaDahale** [comment]
> would you rebase the PR to make it run on CI?

Rebased.

**Comment by berndhahnebach** [comment]
FreeCAD CI-repository:
branch:
https://gitlab.com/berndhahnebach/FreeCAD/-/commits/PR_4985

pipeline:
https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines/376936070

status:
success

all pipelines for each branch: 
https://gitlab.com/berndhahnebach/FreeCAD/-/pipelines?scope=branches

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4985">pipeline status</a>for feature branch [PR_4985](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_4985). Pipeline [#392525122 ](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/392525122) was triggered at [f891e2f](https://github.com/FreeCAD/FreeCAD/pull/4985/commits/f891e2fd6ba70ffad1c4bf1d88b340821a6c38b2). All CI branch [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/Fem/App/FemMesh.cpp [hunk 1]
```diff
@@ -971,25 +971,37 @@ std::set<int> FemMesh::getNodesByEdge(const TopoDS_Edge &edge) const
...
 
+    std::vector<const SMDS_MeshNode*> nodes;
     SMDS_NodeIteratorPtr aNodeIter = myMesh->GetMeshDS()->nodesIterator();
     while (aNodeIter->more()) {
         const SMDS_MeshNode* aNode = aNodeIter->next();
-        Base::Vector3d vec(aNode->X(),aNode->Y(),aNode->Z());
+        nodes.push_back(aNode);
+    }
+
+#pragma omp parallel for schedule(dynamic)
```
Comment (donovaly) [inline src/Mod/Fem/App/FemMesh.cpp:981]: The #pragma tag is new for me. What it is for/how does it work?

**Review by donovaly — COMMENTED** [review]
Looks good.

Please rebase that I can check the compilability.

File: src/Mod/Fem/App/FemMesh.cpp (hunk 1 above)
Comment (AjinkyaDahale) [reply src/Mod/Fem/App/FemMesh.cpp:981]: I used it following the instructions to run openmp, which is for thread level parallelization. The line instructs the program to generate threads as available and have them take care of one iterator of the for loop per thread. Any variables declared inside the loop are local to each thread, but others are shared between all threads.

As for the `#pragma` directive itself, I do not have an exact answer for what it does. The best I could find is that it "is the method specified by the C standard for providing additional information to the compiler, beyond what is conveyed in the language itself" (https://gcc.gnu.org/onlinedocs/cpp/Pragmas.html).

Comment (donovaly) [reply src/Mod/Fem/App/FemMesh.cpp:981]: OK, then please rebase and I will try what the MSVC compiler telly me about this.

Comment (AjinkyaDahale) [reply src/Mod/Fem/App/FemMesh.cpp:981]: Rebased.

**Comment by donovaly** [comment]
Compiles fine and works -> merging

**Pull request merged**